registration/manager_test.go

```diff
@@ -0,0 +1,221 @@
+	id := "cts-123"
+	port := 8558
+	// TODO: update these addresses when /v1/health implemented
+	httpAddress := fmt.Sprintf("http://localhost:%d/v1/status", port)
+	httpsAddress := fmt.Sprintf("https://localhost:%d/v1/status", port)
+	checkID := fmt.Sprintf("%s-health", id)
+
+	testcases := []struct {
+		name     string
+		conf     *RegistrationManagerConfig
+		expected *consulapi.AgentServiceCheck
+	}{
+		{
+			"tls_disabled",
+			&RegistrationManagerConfig{
+				ID:               id,
+				Port:             port,
+				TLSEnabled:       false,
+				SelfRegistration: config.DefaultSelfRegistrationConfig(),
+			},
+			&consulapi.AgentServiceCheck{
+				HTTP:                           httpAddress,
+				Name:                           defaultCheckName,
+				CheckID:                        checkID,
+				Notes:                          defaultCheckNotes,
+				DeregisterCriticalServiceAfter: defaultDeregisterCriticalServiceAfter,
+				Status:                         defaultCheckStatus,
+				Method:                         defaultMethod,
+				Interval:                       defaultInterval,
+				Timeout:                        defaultTimeout,
+				TLSSkipVerify:                  defaultTLSSkipVerify,
+			},
+		},
+		{
+			"tls_enabled",
+			&RegistrationManagerConfig{
+				ID:               id,
+				Port:             port,
+				TLSEnabled:       true,
+				SelfRegistration: config.DefaultSelfRegistrationConfig(),
+			},
+			&consulapi.AgentServiceCheck{
+				HTTP:                           httpsAddress,
+				Name:                           defaultCheckName,
+				CheckID:                        checkID,
+				Notes:                          defaultCheckNotes,
+				DeregisterCriticalServiceAfter: defaultDeregisterCriticalServiceAfter,
+				Status:                         defaultCheckStatus,
+				Method:                         defaultMethod,
+				Interval:                       defaultInterval,
+				Timeout:                        defaultTimeout,
+				TLSSkipVerify:                  defaultTLSSkipVerify,
+			},
+		},
+	}
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			c := defaultHTTPCheck(tc.conf)
+			assert.Equal(t, tc.expected, c)
+		})
+	}
+}
+
+func TestRegistrationManager_RegisterService(t *testing.T) {
+	id := "cts-123"
+	port := 8558
+	ns := "ns-1"
+	check := defaultHTTPCheck(&RegistrationManagerConfig{
+		ID:   id,
+		Port: port,
+		SelfRegistration: &config.SelfRegistrationConfig{
+			Namespace: &ns,
+		},
+	})
+	service := &service{
+		name:      defaultServiceName,
+		tags:      defaultServiceTags,
+		port:      port,
+		id:        id,
+		namespace: ns,
+		checks:    []*consulapi.AgentServiceCheck{check},
+	}
+	testcases := []struct {
+		name      string
+		setup     func(*mocks.ConsulClientInterface)
+		expectErr bool
+	}{
+		{
+			"success",
+			func(cMock *mocks.ConsulClientInterface) {
+				cMock.On("RegisterService", mock.Anything,
+					mock.MatchedBy(func(r *consulapi.AgentServiceRegistration) bool {
```

Comment:
Nice! I didn't know about the MatchedBy function, I'll have to try it out!

Reply:
It's nifty for checking that a request was called with the parameters you expected! I do bang my head a lot when I'm writing them, though, because I tend to get the types wrong (like pointer vs. value) and it's hard to debug why it didn't match. It ends up being a lot of commenting out each attribute check until I figure out which one is messing up!